src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -1,494 +1,380 @@
+        /// <param name="method">the method whose parameteres we are inspecting</param>
+        private void HasConflictingParameterName(MethodDeclarationSyntax method)
         {
-            var hasInvalidParams = method.ParameterList.Parameters.Where(param => SyntaxTokenIs(param.Identifier, "__retval")).Any();
+            // check if the identifier is our special name GeneratedReturnValueName
+            bool IsInvalidParameterName(ParameterSyntax stx) { return stx.Identifier.Value.Equals(GeneratedReturnValueName); }
+
+            var hasInvalidParams = method.ParameterList.Parameters.Where(IsInvalidParameterName).Any();
             if (hasInvalidParams) 
             { 
-                Report(ref context, DiagnosticRules.ParameterNamedValueRule, method.GetLocation(), method.Identifier); 
+                Report(WinRTRules.ParameterNamedValueRule, method.GetLocation(), method.Identifier); 
             }
-            return hasInvalidParams;
         }
 
         /// <summary>
-        ///  Checks to see if an array parameter has been marked with both Write and Read attributes
-        ///  Does extra work, by catching `ref` params, done here since this code can be used by class or interface related methods</summary>
-        /// <param name="method"></param><param name="classIdentifier"></param><param name="context"></param>
-        /// <returns>true if array attributes are invalid (see summary)</returns>
-        private bool CheckParamsForArrayAttributes(MethodDeclarationSyntax method, ref GeneratorExecutionContext context)
+        /// Namespaces can't only differ by cases, also check if the name is invalid for the winmd being made</summary>
+        /// <param name="definedNamespaces"></param><param name="assemblyName">"Test" for unit tests, otherwise stored in AnalyzerconfigPropertyProvider</param>
+        private void HasInvalidNamespace(HashSet<INamespaceSymbol> definedNamespaces, string assemblyName)
         {
-            bool found = false;
-            foreach (ParameterSyntax param in method.ParameterList.Parameters)
-            {
-                var isArrayType = param.ChildNodes().OfType<ArrayTypeSyntax>().Any();
-                bool hasReadOnlyArray = ParamHasAttribute("System.Runtime.InteropServices.WindowsRuntime.ReadOnlyArray", param);
-                bool hasWriteOnlyArray = ParamHasAttribute("System.Runtime.InteropServices.WindowsRuntime.WriteOnlyArray", param);
-                bool isOutputParam = ParamMarkedOutput(param);
+            HashSet<string> simplifiedNames = new HashSet<string>();
 
-                // Nothing can be marked `ref`
-                if (ParamMarkedRef(param))
+            foreach (var namespaceSymbol in definedNamespaces)
+            {
+                string upperNamed = namespaceSymbol.ToString().ToUpper();
+                if (simplifiedNames.Contains(upperNamed))

```

Comment:
What about 2 namespaces defined multiple times with the same case?  Is that a scenario to consider here?

Reply:
If the namespace name is valid (e.g. is the assembly name or a namespace prefix with the assembly name) then that's fine. Otherwise it will get the "disjoint namespace" diagnostic, meaning the types won't be accessible. I'll add a test for this thanks